### ISSUE TITLE
Passing graphs like functions are passed in functional programming 

### ISSUE THREAD

**Issue by seefeldb** [google/breadboard-ai#93]
We currently passing graphs in `include` and `slot`, and that combination if particular powerful: It's almost like passing a function to a function. For example you could imagine a mapOverArray.graph that has a slot for the operation that should happen over each element.

But right now both graphs are self enclosed and not connected to the calling graph. What if we could wire things more or less directly from the calling graph into these passed graphs. It might be a terrible idea, but let's have a look.

Something like `query = "..."; filtered = list.filter(x => x === query)`, where the passed function can access `query` from the closure could become possible.

Two approaches:

## 1) Where we add functions

We could add a way to define a subgraph, but then also wire `query` directly from the context into this subgraph, then pass that subgraph. This could look like this:

```
query = board.input();
node.wire("list->",
  kit.filter()
    .wire("slot<-",
      board.input()
        .wire("x->", kit.compare()
          .wire("y<-.query", query)
          .wire("x->", board.output())))
    .wire()....;
```

(I invented a `kit.compare` node to save typing, that's not part of the suggestion)

It's pretty close to how we'd write it today, except for not serializing to JSON before sending it.

This has implications on google/breadboard-ai#57 (dynamic includes). Also note, re google/breadboard-ai#92, if the wire from query wasn't a constant ('<-.`), this would only work for one round. OTOH, this is how one could write a way to merge streams, though it seems a little shaky...

Note that when serializing this, it's all one graph, with the `filter` node getting a reference to the new `input` node and the `compare` node already wired up with a queued up opportunity. This is a key difference to how `include` and `slot` are implemented: We'd have to reuse the same `GraphRunner` instance, or otherwise have shared queues of opportunities.

Maybe that's how `include` should be implemented even when loading graphs, with the main difference between graphs being their own namespaces for ids and such. This might also eventually help with running things in parallel (though maybe not necessary, we'd just have multiple graph runners running in parallel, with less control on scheduling).

## 2) Using * to pass remainder inputs to subgraphs

This requires less changes. Feels a bit hacky, but maybe it's not: `filter` (whether as actual node or implemented as graph!) just passes all extra incoming wires to its slotted graphs, e.g. here `query`.

```
const innerBoard = new Board();
const compare = kit.compare();

// Side node: I wish this was easier, maybe .wire("[x,query]->" or .wire(["x->", "query->"]?
innerBoard.input().
  .wire("x->", compare)
  .wire("query->", compare)
compare.wire("x->", board.output())

query = board.input();
node.wire("list->",
  kit.filter()
    .wire("query<-", query)
    .wire("slot<-", innerBoard as unknown as NodeValue)
    .wire()....;
```

This makes it impossible for the inner board to output anything outside of `output`, where as above we could add wires out at any point (unlike functional programming!).

All this needs is a way to capture the "remainder" inputs as `*` somehow. Maybe that's not even necessary, it'll just keep passing `list` and `slot` to the inner board, where it gets ignored. But this makes the issue with this approach more apparent: There is potential for name collisions, especially if the operators (map, filter, etc) are being developed independently of what can go inside.

### 2b) bundler nodes?

To address that last issue and create a namespace we could create bundler nodes that either collect `*` into a single value or expand such a single value into a `*`. 

So
```
...
   .wire("slotArgs<-.bundle", board.bundle().wire("<-query", query))
```
in the caller (with a convention that matches up slot names with arguments)

And
```
...
   .wire("<-query", board.unbundle().wire("bundle<-args").input)
...
```
in the function (with a convention that this is passed as `args`).

A bit more typing, but could work.

Alternatively, skip that second change in the function and do pass all args expanded already. The namespace issue is really more in the operator and the bundle tunnels through that. Then it's just a few extra characters to bundle the inputs.

## 3) `bind` as a mix of the two, and probably the best option

Here we bind a set of inputs to a graph representation (or board) and then pass that as slot input. Once the runner gets to this board, it adds these inputs to the ones passed in by the caller with the caller inputs getting precedence.

Syntax TBD.

FWIW:

A nicer syntax for either might be to have kit.map() accept a (JavaScript) function that gets board.input() and board.output() as parameters, with the slot made implicit:

```
query = board.input();
node.wire("list->",
  kit.filter((in, out) => in.wire("x->", kit.compare()
          .wire("y<-.query", query) // or `in` for the second proposal
          .wire("x->", out))
    .wire()....;
```

or something like that to make it all feel more like regular javascript code.
